src/List.ts

```diff
@@ -0,0 +1,106 @@
+import { Iterable, IterableIterator, ShimIterator } from 'dojo-shim/iterator';
+
+export interface List<T> extends Iterable<T> {
+	readonly size: number;
+
+	add(value: T): this;
+	clear(): void;
+	delete(idx: number): boolean;
+	entries(): IterableIterator<[ number, T ]>;
+	forEach(fn: (value: T, idx: number, list: this) => void, thisArg?: any): void;
+	has(idx: number): boolean;
+	includes(value: T): boolean;
+	indexOf(value: T): number;
+	join(separator?: string): string;
+	keys(): IterableIterator<number>;
+	lastIndexOf(value: T): number;
+	pop(): T | undefined;
+	push(value: T): void;
+	splice(start: number, deleteCount?: number, ...newItems: T[]): T[];
+	values(): IterableIterator<T>;
+}
+
+export default class ArrayList<T> implements List<T> {
+	private _items: T[];
```

Comment:
I think we need to avoid private members on classes.  First, it isn't truly private, second, the `private` keyword won't be the way it gets implemented in ECMAScript and third, it causes all sorts of assignability issues when used in TypeScript.  Even though it is sort of "meta" I would use a weak map to store the items.  This is how they are suggesting people will eventually polyfill the private class members anyways.